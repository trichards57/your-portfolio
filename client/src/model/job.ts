export type Outcome =
  | "DischargedOnScene"
  | "NotFound"
  | "StoodDown"
  | "Conveyed"
  | "Other";

export interface NewJob {
  age?: number;
  blueLights: boolean;
  category: 1 | 2 | 3 | 4 | 5;
  drove: boolean;
  gender?: "Male" | "Female";
  notes: string;
  outcome: Outcome;
  reflectionFlag: boolean;
  shift: string;
}

export interface JobSummary {
  age?: number;
  category: 1 | 2 | 3 | 4 | 5;
  gender?: "Male" | "Female";
  reflectionFlag: boolean;
}
