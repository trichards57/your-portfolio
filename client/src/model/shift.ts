export type RoleType = "EAC" | "AFA" | "CRU";

export interface Shift {
  crewMate?: string;
  date: string;
  duration: number;
  event: string;
  id: string;
  location?: string;
  role: RoleType;
}

export interface ShiftSummary extends Shift {
  loggedCalls: number;
}

export type NewShift = Omit<Shift, "id">;
