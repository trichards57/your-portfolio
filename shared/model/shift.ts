export interface ShiftSummary {
  id: number;
  date: string;
  event: string;
  location: string;
  duration: number;
  role: "EAC" | "AFA" | "CRU";
  crewMate: string;
  loggedCalls: number;
}
