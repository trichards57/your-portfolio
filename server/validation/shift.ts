import { isValid, parseISO } from "date-fns";
import { ValidationError } from "./validation-error";
import { NewShift } from "../../shared/model/shift";

export function validateNewShift(
  item: Partial<NewShift>
): [ValidationError<NewShift>[], NewShift | undefined] {
  let res: ValidationError<NewShift>[] = [];

  if (!item.date || !isValid(parseISO(item.date)))
    res.push({
      error: "invalid date",
      fieldName: "date",
    });

  if (!item.event || item.event.trim() === "")
    res.push({
      error: "missing event name",
      fieldName: "event",
    });

  if (!(item.role == "AFA" || item.role == "CRU" || item.role == "EAC"))
    res.push({
      error: "invalid role",
      fieldName: "role",
    });

  if (typeof item.duration != "number" || !Number.isFinite(item.duration))
    res.push({
      error: "invalid duration",
      fieldName: "duration",
    });

  return [res, item as NewShift];
}
