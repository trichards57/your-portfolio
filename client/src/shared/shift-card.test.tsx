import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter as Router } from "react-router-dom";
import ShiftCard from "./shift-card";
import { ShiftSummary } from "../model/shift";
import renderer from "react-test-renderer";

const defaultShift: ShiftSummary = {
  date: "2021-10-01",
  duration: 5,
  event: "SWASFT Shift",
  id: "abcdef",
  loggedCalls: 5,
  role: "EAC",
  crewMate: "Test Crewmate",
  location: "Test Location",
};

it("renders without crashing", () => {
  const div = document.createElement("div");

  ReactDOM.render(
    <Router>
      <ShiftCard shift={defaultShift} />
    </Router>,
    div
  );

  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Router>
        <ShiftCard shift={defaultShift} />
      </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
