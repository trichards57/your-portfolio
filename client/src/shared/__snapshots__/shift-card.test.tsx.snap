// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardContent-root"
  >
    <p
      className="MuiTypography-root makeStyles-date-4 makeStyles-trimText-5 MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-gutterBottom"
    >
      1 October 2021
    </p>
    <h2
      className="MuiTypography-root makeStyles-trimText-5 MuiTypography-h5"
    >
      SWASFT Shift
    </h2>
    <p
      className="MuiTypography-root makeStyles-location-6 makeStyles-trimText-5 MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Test Location
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2"
    >
      Role: 
      EAC
    </p>
    <p
      className="MuiTypography-root MuiTypography-body2"
    >
      Logged Calls: 
      5
    </p>
  </div>
  <div
    className="MuiCardActions-root MuiCardActions-spacing"
  >
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
      href="/editShift/abcdef"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiButton-label"
      >
        Edit
      </span>
    </a>
    <a
      aria-disabled={false}
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
      href="/addJob?shift=abcdef"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiButton-label"
      >
        Add Job
      </span>
    </a>
  </div>
</div>
`;
